Xamarin.Forms.Platform.Android/Forms.cs

```diff
@@ -191,54 +193,149 @@ static Color GetAccentColor()
+				SetScreenOrientation();
+				_formsActivity.ConfigurationChanged += OnConfigurationChanged;
 			}
 
-			public override Size PixelScreenSize
+			public override Size PixelScreenSize => _pixelScreenSize;
+
+			public override Size ScaledScreenSize { get; }
+
+			public override double ScalingFactor => _scalingFactor;
+
+			public override void BeginDeviceOrientationNotifications()
 			{
-				get { return _pixelScreenSize; }
+				if (_androidOrientationEventListener != null)
+					return;
+
+				_androidOrientationEventListener = new AndroidOrientationEventListener((Context)_formsActivity, SensorDelay.Normal);
+				_androidOrientationEventListener.OrientationChanged += OnDeviceOrientationChanged;
+
+				if (_androidOrientationEventListener.CanDetectOrientation())
+					_androidOrientationEventListener.Enable();
 			}
 
-			public override Size ScaledScreenSize { get; }
+			public override void EndDeviceOrientationNotifications()
+			{
+				if (_androidOrientationEventListener == null)
+				    return;
+
+				_androidOrientationEventListener.OrientationChanged -= OnDeviceOrientationChanged;
+				_androidOrientationEventListener.Disable();
+				_androidOrientationEventListener.Dispose();
+				_androidOrientationEventListener = null;
+			}
+
+			void OnDeviceOrientationChanged(object sender, int i)
+			{
+				SetDeviceOrientation(i);
+			}
+
+			void OnConfigurationChanged(object sender, EventArgs e)
+			{
+				SetScreenOrientation();
+			}
 
-			public override double ScalingFactor
+			void SetDeviceOrientation(int rotation)
 			{
-				get { return _scalingFactor; }
+				if (rotation == OrientationEventListener.OrientationUnknown)
+				{
+					DeviceOrientation = DeviceOrientation.Unknown;
+					return;
+				}
+
+				if (PixelScreenSize.Width < PixelScreenSize.Height)
+				{
+					const int threshold = 45;
+
+					if (rotation <= threshold || rotation > 360 - threshold)
+						DeviceOrientation = DeviceOrientation.Portrait;
+					else if (rotation <= 360 - threshold && rotation > 270 - threshold)
+						DeviceOrientation = DeviceOrientation.Landscape;
+					else if (rotation <= 270 - threshold && rotation > 180 - threshold)
+						DeviceOrientation = DeviceOrientation.PortraitFlipped;
+					else if (rotation <= 180 - threshold && rotation > threshold)
+						DeviceOrientation = DeviceOrientation.LandscapeFlipped;
+				}
+				else if (PixelScreenSize.Width > PixelScreenSize.Height)
+				{
+					const int threshold = 45;
+
+					if (rotation <= threshold || rotation > 360 - threshold)
+						DeviceOrientation = DeviceOrientation.Landscape;
+					else if (rotation <= 360 - threshold && rotation > 270 - threshold)
+						DeviceOrientation = DeviceOrientation.PortraitFlipped;
+					else if (rotation <= 270 - threshold && rotation > 180 - threshold)
+						DeviceOrientation = DeviceOrientation.LandscapeFlipped;
+					else if (rotation <= 180 - threshold && rotation > threshold)
+						DeviceOrientation = DeviceOrientation.Portrait;
+				}
```

Comment:
I think I have an idea how to go about further improving this:

`DeviceOrientation`: orientation of the physical device as reported by a sensor manager.
`ScreenOrientation`: orientation of the display monitor/interface
`LayoutOrientation`: orientation of your UI content

See https://api.jquerymobile.com/orientationchange/. On Windows, you can resize your browser window with the grab handles so that JQuery thinks you're in landscape or portrait, but your screen orientation has not changed (right click on desktop on Win7 -> Screen resolution -> Orientation). Your device orientation has not changed, but your desktop most likely does not have a built-in sensor (even most laptops don't except for Surface Book).

In all cases, if an enum is not supported, it should default to `Unknown`. There should be strong API documentation to let devs be aware of environmental differences.